ipa-core/src/protocol/hybrid/agg.rs

```diff
@@ -0,0 +1,235 @@
+        BooleanProtocols, RecordId,
+    },
+    report::hybrid::{AggregateableHybridReport, PrfHybridReport},
+    secret_sharing::{replicated::semi_honest::AdditiveShare as Replicated, SharedValue},
+    seq_join::{seq_join, SeqJoin},
+};
+
+/// This protocol is used to aggregate `PRFHybridReports` and returns `AggregateableHybridReports`.
+/// It groups all the reports by the PRF of the `match_key`, finds all reports from `match_keys`
+/// with that provided exactly 2 reports, then adds those 2 reports.
+pub async fn aggregate_reports<BK, V, C>(
+    ctx: C,
+    reports: Vec<PrfHybridReport<BK, V>>,
+) -> Result<Vec<AggregateableHybridReport<BK, V>>, Error>
+where
+    C: UpgradableContext,
+    BK: SharedValue + BooleanArray,
+    V: SharedValue + BooleanArray,
+    Replicated<Boolean>: BooleanProtocols<DZKPUpgraded<C>>,
+{
+    let mut reports_by_matchkey = HashMap::with_capacity(reports.len() / 2);
+
+    // build a hashmap of match_key -> ([AggregateableHybridReport;2], count)
+    // if count ever exceeds 2, we drop reports, but keep counting
+    // an honest client and report collector will only submit
+    // one report with a breakdown key and one report with a value.
+    // if there are less, it's unattributed. if more, something went wrong.
+    for report in reports {
+        let match_key = report.match_key;
+        let entry = reports_by_matchkey.entry(match_key).or_insert((
+            [
+                AggregateableHybridReport::<BK, V>::ZERO,
+                AggregateableHybridReport::<BK, V>::ZERO,
+            ],
+            0,
+        ));
+        if entry.1 == 0 {
+            // If the count is 0, replace the first element with the new report
+            entry.0[0] = report.into();
+            entry.1 += 1;
+        } else if entry.1 == 1 {
+            // If the count is 1, replace the second element with the new report
+            entry.0[1] = report.into();
+            entry.1 += 1;
+        } else {
+            // If the count is 2 or more, increment the counter and drop the report
+            entry.1 += 1;
+        }
+    }
+
+    // we only keep the reports from match_keys that provided exactly 2 reports
+    let report_pairs: Vec<[AggregateableHybridReport<BK, V>; 2]> = reports_by_matchkey
+        .into_iter()
+        .filter_map(|(_, v)| if v.1 == 2 { Some(v.0) } else { None })
+        .collect::<Vec<_>>();
+
+    // let mut agg_reports = Vec::with_capacity(report_pairs.len());
+
+    let chunk_size = TARGET_PROOF_SIZE;
+
+    let dzkp_validator = ctx.clone().dzkp_validator(
+        MaliciousProtocolSteps {
+            protocol: &HybridStep::GroupBySum,
+            validate: &HybridStep::GroupBySumValidate,
+        },
+        std::cmp::min(ctx.active_work().get(), chunk_size.next_power_of_two()),
+    );
+
+    let ctx = dzkp_validator
+        .context()
+        .set_total_records(TotalRecords::specified(2 * report_pairs.len())?);
```

Comment:
I did `2 * report_pairs.len()` here because we do 2 additions per row in the vec, but if we're just using the index in the vec as the record_id, should this not be 2x? cc @akoshelev

Reply:
yes, it should **not** be 2x I think